### PR TITLE
[AutoBuild] new feature: function calling

### LINKED ISSUE #1510: [Feature Request]: how do we add tools to agents created by AgentBuilder ?
### Is your feature request related to a problem? Please describe.

I want to integrate my custom functions and langchain tools with agents created by AgentBuilder

### Describe the solution you'd like

It should be possible to use custom functions or langchain's tools by these agents, As other UserProxyAgent, GPTAssistantAgent and many have the option to. If this is a feature on which work needs to be started, I can also take it up 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by krishnashed**
## Why are these changes needed?

Introduction of Autobuild, didn't support function/tool calling. This PR is solution to add this feature/

## Related issue number

Closes #1510 

## Checks
Check and tests still need to be written and passed.

### CODE REVIEW

**Comment by LinxinS97** [comment]
Autobuild will dramatically change in the next few weeks with newly designed prompts and a new retrieval system, including how to bind tools with agents. I suggest closing this PR and resubmitting it after the new Autobuild release.

**Comment by sonichi** [comment]
@LinxinS97 thanks for the headsup. Could you include @krishnashed as a reviewer for the changes? If there is a mutual interest, also invite @krishnashed to test it.

**Comment by krishnashed** [comment]
Sure @sonichi @LinxinS97 I'm very interested, I would like to contribute in any way I can.

**Comment by krishnashed** [comment]
> Autobuild will dramatically change in the next few weeks with newly designed prompts and a new retrieval system, including how to bind tools with agents. I suggest closing this PR and resubmitting it after the new Autobuild release.

@LinxinS97 I looked up open issues/PRs but didn't find anything revolving around change in tools/function bindings, Could you share some links about where we are heading to? Is there something I can contribute to in that work ?

Is this the format we will define and utilize functions: https://microsoft.github.io/autogen/docs/topics/code-execution/user-defined-functions ? So should the changes be made according to this ?

**Comment by LinxinS97** [comment]
> > Autobuild will dramatically change in the next few weeks with newly designed prompts and a new retrieval system, including how to bind tools with agents. I suggest closing this PR and resubmitting it after the new Autobuild release.
> 
> @LinxinS97 I looked up open issues/PRs but didn't find anything revolving around change in tools/function bindings, Could you share some links about where we are heading to? Is there something I can contribute to in that work ?
> 
> Is this the format we will define and utilize functions: https://microsoft.github.io/autogen/docs/topics/code-execution/user-defined-functions ? So should the changes be made according to this ?

We will separate `tool_builder` and `agent_builder` into two files. `tool_builder` will be released next week, and you can continue contributing to that PR.

**Comment by krishnashed** [comment]
Sure @LinxinS97 that's great. Can you please point me to the PR, GitHub Issue or a doc where it's described more in detail. So that I can get started ?

**Comment by LinxinS97** [comment]
Let me know after you complete the following steps:

- Merge your PR with the latest autobuild.
- Change your notebook into a step-by-step format; refer to `notebook/autobuild_basic.ipynb`.
- Provide a successful sample in your notebook with the corresponding output.

**Comment by krishnashed** [comment]
@LinxinS97  @JieyuZ2  @sonichi  this PR is ready for review! Im not sure why the llamaindex checks are failing tho, how can i fix them ?

File: notebook/autobuild_function_calling.ipynb [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+    "config_file_or_env = \"OAI_CONFIG_LIST\"\n",
+    "config_list = autogen.config_list_from_json(config_file_or_env, filter_dict={\"model\": [\"gpt-4-1106-preview\", \"gpt-4\"]})\n",
+    "llm_config = {\n",
+    "    \"config_list\": config_list,\n",
+    "    \"timeout\": 120,\n",
+    "}\n",
+    "\n",
+    "\n",
+    "def start_task(execution_task: str, agent_list: list):\n",
+    "    group_chat = autogen.GroupChat(agents=agent_list, messages=[], max_round=12)\n",
```
Comment (LinxinS97) [inline notebook/autobuild_function_calling.ipynb:58]: You can prevent User_console from being called multiple times by adding `allow_repeat_speaker=[list of agent except user_proxy]` when initializing group chat.

**Comment by LinxinS97** [comment]
The example in the notebook needs to be better because the model cannot load and read the image, which causes inconsistency in conversation. You might consider using another example to exhibit this feature better.

File: notebook/autobuild_function_calling.ipynb [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+   "outputs": [],
+   "source": [
+    "import time\n",
+    "from datetime import datetime, timedelta\n",
+    "from typing import Dict\n",
+    "\n",
+    "import feedparser\n",
+    "\n",
+    "\n",
+    "def get_arxiv_papers() -> Dict:\n",
```
Comment (LinxinS97) [inline notebook/autobuild_function_calling.ipynb:123]: The tool's name should reflect its features. `get_arxiv_paper_from_yesterday` is better.

File: notebook/autobuild_function_calling.ipynb [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    "    # arXiv API endpoint\n",
+    "    url = \"http://export.arxiv.org/api/query?\"\n",
+    "\n",
+    "    # Search parameters\n",
+    "    max_results = 10\n",
```
Comment (LinxinS97) [inline notebook/autobuild_function_calling.ipynb:128]: This should be a function parameter. 
Also, you should add some arguments for accurate searches, like `research_field` or `research_domain`, and wrap them into the search query by your tool. Imagine how you use a search engine.

File: notebook/autobuild_function_calling.ipynb [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[33mArxivAPI_Expert\u001b[0m (to chat_manager):\n",
+      "\n",
+      "Get arxiv papers published in last 24 hours\n",
```
Comment (LinxinS97) [inline notebook/autobuild_function_calling.ipynb:242]: The task should reflect the collaboration between agents, such as "analysis papers in the field of computer science in the last 24 hours."

**Review by LinxinS97 — APPROVED** [review]

**Pull request closed without merging**